libs/Wallet/src/NewWalletAccountController.cpp

```diff
@@ -0,0 +1,188 @@
+#include <StatusGo/Types.h>
+
+#include <Onboarding/Common/Constants.h>
+
+#include <QQmlEngine>
+#include <QJSEngine>
+
+#include <ranges>
+
+namespace GoAccounts = Status::StatusGo::Accounts;
+namespace GoWallet = Status::StatusGo::Wallet;
+namespace UtilsSG = Status::StatusGo::Utils;
+namespace StatusGo = Status::StatusGo;
+
+namespace Status::Wallet {
+
+NewWalletAccountController::NewWalletAccountController(std::shared_ptr<Helpers::QObjectVectorModel<WalletAccount>> accounts)
+    : m_accounts(accounts)
+    , m_mainAccounts(std::move(filterMainAccounts(*accounts)), "account")
+    , m_derivedAddress("derivedAddress")
+    , m_derivationPath(Status::Constants::General::PathWalletRoot)
+{
+}
+
+NewWalletAccountController::~NewWalletAccountController()
+{
+}
+
+QAbstractListModel* NewWalletAccountController::mainAccountsModel()
+{
+    return &m_mainAccounts;
+}
+
+QAbstractItemModel *NewWalletAccountController::currentDerivedAddressModel()
+{
+    return &m_derivedAddress;
+}
+
+QString NewWalletAccountController::derivationPath() const
+{
+    return m_derivationPath.get();
+}
+
+void NewWalletAccountController::setDerivationPath(const QString &newDerivationPath)
+{
+    if (m_derivationPath.get() == newDerivationPath)
+        return;
+    m_derivationPath = GoAccounts::DerivationPath(newDerivationPath);
+    emit derivationPathChanged();
+
+    auto oldCustom = m_customDerivationPath;
+    auto found = searchDerivationPath(m_derivationPath);
+    m_customDerivationPath = std::get<0>(found) == nullptr;
+    if(!m_customDerivationPath && !std::get<0>(found).get()->alreadyCreated())
+        updateSelectedDerivedAddress(std::get<1>(found), std::get<0>(found));
+
+    if(m_customDerivationPath != oldCustom)
+        emit customDerivationPathChanged();
+}
+
+void NewWalletAccountController::createAccountAsync(const QString &password, const QString &name,
+                                          const QColor &color, const QString &path,
+                                          const WalletAccount *derivedFrom)
+{
+    try {
+        GoAccounts::generateAccountWithDerivedPath(StatusGo::HashedPassword(UtilsSG::hashPassword(password)),
+                                                   name, color, "", GoAccounts::DerivationPath(path),
+                                                   derivedFrom->data().derivedFrom.value());
+        auto found = findMissingAccount();
+        if(found)
+            m_accounts->push_back(found);
+        else
+            qWarning() << "Failed to create account. No new account found by this->findMissingAccount";
+
+        emit accountCreatedStatus(found != nullptr);
+    }
+    catch(const StatusGo::CallPrivateRpcError& e) {
+        qWarning() << "StatusGoQt.generateAccountWithDerivedPath error: " << e.errorResponse().error.message.c_str();
+        emit accountCreatedStatus(false);
+    }
+}
+
+bool NewWalletAccountController::retrieveAndUpdateDerivedAddresses(const QString &password,
```

Comment:
This function could use comments, because even if you know how multiwallets work, it's still not easy to understand.

Reply:
@jrainville, thanks for the heads up regarding the confusion. Here is the story behind this:

> Also, maybe it's just me, but ChatOrWalletAccount is kinda confusing.

`ChatOrWalletAccount` is the equivalent of the `WalletAccountDto` in NIM which confused me to looking in the wrong direction. I tried to make it more explicit by renaming it into what it is :).

> I know why you chose that name, but still, I personally like MultiAccount more, but it's probably because I'm used to it?

- I already have a [`MultiAccount`](https://github.com/status-im/status-desktop/blob/aa1d4dabee6adeb7142453398a1de2c43817dafb/libs/Onboarding/src/Onboarding/Accounts/MultiAccount.h#L18) entity here in the Onboarding module, with the TODO to move it into the `StatusGoQt` warapper, under accounts, next. This is the equivalent of `AccountDto` in NIM which I found to be the actual identifier of the Status multi-account concept.

All the above are status-go concepts that are next mapped to the presentation layer entities like `UserAccount` (for multi-account) and `WalletAccount`/`ChatAccount` entities that are more close to the user workflow.

> I think having some comments on the more complex functions could help future people understanding what is going on.

I will try to improve; thanks for bringing it to my attention.